src/EmailGuardianVerifier.sol

```diff
@@ -0,0 +1,553 @@
+                    emailData.commandParams
+                );
+
+            require(
+                account == recoveredAccount,
+                "invalid account in acceptance command"
+            );
+
+            templateId = computeAcceptanceTemplateId(emailData.templateIdx);
+        }
+
+        require(emailData.isCodeExist == true, "isCodeExist is false");
+
+        EmailProof memory emailProof = EmailProof({
+            domainName: emailData.domainName,
+            publicKeyHash: proof.publicInputs[0],
+            timestamp: emailData.timestamp,
+            maskedCommand: emailData.maskedCommand,
+            emailNullifier: proof.publicInputs[1],
+            accountSalt: emailData.accountSalt,
+            isCodeExist: emailData.isCodeExist,
+            proof: proof.proof
+        });
+
+        string[] memory template = commandTemplates[templateId];
+
+        require(template.length != 0, "invalid template id");
+
+        require(
+            dkimRegistry.isDKIMPublicKeyHashValid(
+                emailProof.domainName,
+                emailProof.publicKeyHash
+            ) == true,
+            "invalid dkim public key hash"
+        );
+
+        require(
+            usedNullifiers[emailProof.emailNullifier] == false,
+            "email nullifier already used"
+        );
+
+        // TODO: match with the accountSalt added during initialization
+        // require(
+        //     accountSalt == emailAuthMsg.proof.accountSalt,
+        //     "invalid account salt"
+        // );
+
+        require(
+            timestampCheckEnabled == false ||
+                emailProof.timestamp == 0 ||
+                emailProof.timestamp > lastTimestamp,
+            "invalid timestamp"
+        );
+
+        require(
+            bytes(emailProof.maskedCommand).length <= verifier.commandBytes(),
+            "invalid masked command length"
+        );
+
+        require(
+            emailData.skippedCommandPrefix < verifier.commandBytes(),
+            "invalid size of the skipped command prefix"
+        );
+
+        string memory trimmedMaskedCommand = removePrefix(
+            emailProof.maskedCommand,
+            emailData.skippedCommandPrefix
+        );
+        string memory expectedCommand = "";
+        for (uint stringCase = 0; stringCase < 3; stringCase++) {
+            expectedCommand = CommandUtils.computeExpectedCommand(
+                emailData.commandParams,
+                template,
+                stringCase
+            );
+            if (Strings.equal(expectedCommand, trimmedMaskedCommand)) {
+                break;
+            }
+            if (stringCase == 2) {
+                revert("invalid command");
+            }
+        }
+
+        require(
+            verifier.verifyEmailProof(emailProof) == true,
+            "invalid email proof"
+        );
+
+        // TODO: How to handle the nullifier update
+        // usedNullifiers[emailProof.emailNullifier] = true;
```

Comment:
ok conclusion here is that we can do replay protection outside of these functions. I'm not 100% convinced of the right design here but this could either be in the recovery module itself, or in a wrapper function in the guardian verifier.

Kind of an unhelpful and unspecific answer but haven't had time to deeply think through this.

Default option could be to do it in the recovery module. If a nullifier is not in all the proofs then you might have to build some more generalised logic to handle this. If all verifiers use nullifiers then you can just use that.

Reply:
If nullifier is added to records in a separate function ( as verifyProof is a view one ) then I think it defies the purpose of having nullifier.

I like the idea of adding it in recovery module, as it becomes generalised then and we can ensure that no proof is reused for recovery/acceptance